### PR TITLE
bpo-36737: Use the module state C-API for warnings.

### PR DESCRIPTION
**Pull request by ericsnowcurrently**
https://bugs.python.org/issue36737

### CODE REVIEW

File: Python/_warnings.c [hunk 1]
```diff
@@ -15,6 +15,109 @@ _Py_IDENTIFIER(default);
 _Py_IDENTIFIER(ignore);
 #endif
 
+
+/*************************************************************************/
+
+typedef struct warnings_state {
```
Comment (zooba) [inline Python/_warnings.c]: Why is this so inconsistent with everything else? (`WarningsState`, `struct warningsmodule`, etc.)

File: Python/_warnings.c [hunk 2]
```diff
@@ -1255,50 +1352,38 @@ _PyWarnings_Init(void)
 {
     PyObject *m;
 
+    //m = PyState_FindModule(&warningsmodule);
```
Comment (zooba) [inline Python/_warnings.c]: Remove?

File: Python/_warnings.c [hunk 3]
```diff
@@ -1255,50 +1352,38 @@ _PyWarnings_Init(void)
...
-        return NULL;
 
-    if (state->once_registry == NULL) {
-        state->once_registry = PyDict_New();
-        if (state->once_registry == NULL)
-            return NULL;
+    WarningsState *st = _Warnings_GetState(m);
+
+    if (_Warnings_InitState(st) < 0) {
+        // XXX decref m?
```
Comment (zooba) [inline Python/_warnings.c]: Yes

```suggestion
        Py_DECREF(m);
```

File: Python/_warnings.c [hunk 4]
```diff
@@ -1255,50 +1352,38 @@ _PyWarnings_Init(void)
...
-    if (PyModule_AddObject(m, "filters", state->filters) < 0)
-        return NULL;
 
-    if (state->once_registry == NULL) {
-        state->once_registry = PyDict_New();
-        if (state->once_registry == NULL)
-            return NULL;
+    WarningsState *st = _Warnings_GetState(m);
+
+    if (_Warnings_InitState(st) < 0) {
```
Comment (zooba) [inline Python/_warnings.c:1349]:
```suggestion
    if (st == NULL || _Warnings_InitState(st) < 0) {
```

File: Python/_warnings.c (hunk 3 above)
Comment (ericsnowcurrently) [reply Python/_warnings.c]: fixed

File: Python/_warnings.c (hunk 2 above)
Comment (ericsnowcurrently) [reply Python/_warnings.c]: done

File: Python/_warnings.c (hunk 1 above)
Comment (ericsnowcurrently) [reply Python/_warnings.c]: fixed

File: Python/_warnings.c (hunk 4 above)
Comment (ericsnowcurrently) [reply Python/_warnings.c:1349]: fixed

**Review by zooba — APPROVED** [review]

**Pull request merged**